src/hotspot/share/classfile/systemDictionary.cpp

```diff
@@ -368,7 +368,7 @@ Klass* SystemDictionary::resolve_array_class_or_null(Symbol* class_name,
     }
   } else {
     k = Universe::typeArrayKlass(t);
-    k = TypeArrayKlass::cast(k)->array_klass(ndims, CHECK_NULL);
+    k = k->array_klass(ndims, CHECK_NULL);
```

Comment:
I assume the cast was an attempt to de-virtualize the array_klass() call, so it is better not to use Klass* here.

Reply:
My experience is that these type of casts doesn't make the compiler devirtualize the calls. I tried it now and verified that both with and without the cast we still get the virtual call. You typically need to tell the compiler what function it should be using. (I played around a lot with this when writing the devirtualization layer for the oop_iterate/OopIterateClosure code.)

I tested writing the code above as `TypeArrayKlass::cast(k)->TypeArrayKlass::array_klass(ndims, CHECK_NULL)` and that gets rid of the virtual call. However, the compiler still can't inline the code ArrayKlass::array_klass code because it is inside a .cpp file and not an .inline.hpp, so this results in a direct call instead of inlined code.

Reply:
OK, I guess the compiler needs to be conservative in case TypeArrayKlass has a subclass.